examples/server/server.cpp

```diff
@@ -2016,6 +2037,29 @@ struct server_context {
                                     {"prompt_tokens",   tokens_to_str(ctx, prompt_tokens.cbegin(), prompt_tokens.cend())},
                                 });
 
+                                if (n_truncate && slot.params.cache_prompt) {
+                                    const int n_keep = slot.params.n_keep + add_bos_token;
+                                    for (size_t i = n_keep + n_truncate; i < slot.cache_tokens.size(); i++) {
+                                        slot.cache_tokens[i - n_truncate] = slot.cache_tokens[i];
+                                    }
+
+                                    const int new_cache_size = slot.cache_tokens.size() - n_truncate;
+                                    if (new_cache_size >= 0) {
+                                        slot.cache_tokens.resize(new_cache_size);
+
+                                        LOG_VERBOSE("cache tokens shift", {
+                                            {"id_slot",        slot.id},
+                                            {"id_task",        slot.id_task},
+                                            {"n_ctx",          slot.n_ctx},
+                                            {"n_keep",         slot.params.n_keep},
+                                            {"n_left",         n_left},
+                                            {"n_truncate",     n_truncate},
+                                            {"new_cache_size", new_cache_size},
+                                            {"cache_tokens",   tokens_to_str(ctx, slot.cache_tokens.cbegin(), slot.cache_tokens.cend())},
+                                        });
+                                    } // else somebody trying to use n_truncate w/o previous cache
```

Comment:
Is this a forgotten TODO? What should be happening in this case?

Reply:
I'm not so sure about this one, maybe just log an error, using n_truncate only makes sense when you have a cache, otherwise you could just truncate the prompt yourself.